ros/src/computing/planning/mission/packages/op_global_planner/nodes/op_global_planner_core.cpp

```diff
@@ -0,0 +1,392 @@
+ */
+
+#include "op_global_planner_core.h"
+#include "op_RosHelpers.h"
+
+namespace GlobalPlanningNS
+{
+
+GlobalPlanner::GlobalPlanner()
+{
+	m_pCurrGoal = 0;
+	m_iCurrentGoalIndex = 0;
+	m_bKmlMap = false;
+	m_bFirstStart = false;
+
+	nh.getParam("/op_global_planner/pathDensity" 			, m_params.pathDensity);
+	nh.getParam("/op_global_planner/enableSmoothing" 		, m_params.bEnableSmoothing);
+	nh.getParam("/op_global_planner/enableLaneChange" 	, m_params.bEnableLaneChange);
+	nh.getParam("/op_global_planner/enableRvizInput" 		, m_params.bEnableRvizInput);
+	nh.getParam("/op_global_planner/enableReplan" 		, m_params.bEnableReplanning);
+	nh.getParam("/op_global_planner/mapFileName" 			, m_params.KmlMapPath);
+
+	int iSource = 0;
+	nh.getParam("/op_global_planner/mapSource" 			, iSource);
+	if(iSource == 0)
+		m_params.mapSource = PlannerHNS::MAP_AUTOWARE;
+	else if (iSource == 1)
+		m_params.mapSource = PlannerHNS::MAP_FOLDER;
+	else if(iSource == 2)
+		m_params.mapSource = PlannerHNS::MAP_KML_FILE;
+
+
+	tf::StampedTransform transform;
+	PlannerHNS::RosHelpers::GetTransformFromTF("map", "world", transform);
+	m_OriginPos.position.x  = transform.getOrigin().x();
+	m_OriginPos.position.y  = transform.getOrigin().y();
+	m_OriginPos.position.z  = transform.getOrigin().z();
+
+	pub_Paths = nh.advertise<autoware_msgs::LaneArray>("lane_waypoints_array", 1, true);
+	pub_PathsRviz = nh.advertise<visualization_msgs::MarkerArray>("global_waypoints_rviz", 1, true);
+	pub_StartPointRviz = nh.advertise<visualization_msgs::Marker>("Global_StartPoint_rviz", 1, true);
+	pub_GoalPointRviz = nh.advertise<visualization_msgs::MarkerArray>("Global_GoalPoints_rviz", 1, true);
+	pub_NodesListRviz = nh.advertise<visualization_msgs::MarkerArray>("Goal_Nodes_Points_rviz", 1, true);
+	pub_MapRviz  = nh.advertise<visualization_msgs::MarkerArray>("vector_map_center_lines_rviz", 100, true);
+	pub_TrafficInfoRviz = nh.advertise<visualization_msgs::MarkerArray>("Traffic_Lights_rviz", 1, true);
+	pub_GoalsListRviz = nh.advertise<visualization_msgs::MarkerArray>("op_destinations_rviz", 1, true);
+
+	if(m_params.bEnableRvizInput)
+	{
+		sub_start_pose 	= nh.subscribe("/initialpose", 					1, &GlobalPlanner::callbackGetStartPose, 		this);
+		sub_goal_pose 	= nh.subscribe("move_base_simple/goal", 		1, &GlobalPlanner::callbackGetGoalPose, 		this);
+	}
+	else
+	{
+		LoadSimulationData();
+	}
+
+	sub_current_pose 		= nh.subscribe("/current_pose", 			100,	&GlobalPlanner::callbackGetCurrentPose, 		this);
+
+	int bVelSource = 1;
+	nh.getParam("/op_global_planner/velocitySource", bVelSource);
+	if(bVelSource == 0)
+		sub_robot_odom 			= nh.subscribe("/odom", 					100,	&GlobalPlanner::callbackGetRobotOdom, 	this);
+	else if(bVelSource == 1)
+		sub_current_velocity 	= nh.subscribe("/current_velocity",		100,	&GlobalPlanner::callbackGetVehicleStatus, 	this);
+	else if(bVelSource == 2)
+		sub_can_info 			= nh.subscribe("/can_info",		100,	&GlobalPlanner::callbackGetCanInfo, 	this);
```

Comment:
fomatting of the code is messed up

Reply:
Otherwise also not the best practice to hard code topics

Reply:
can you share link with the best practice of topic definition

Reply:
no magic. set a param in launch file and then read this param in above program. you can still keep the default value.